File: 130-surrounded-regions/surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
        int m = board.length;
        int n = board[0].length;

        boolean[][] vis = new boolean[m][n];

        //1st row
        for(int j = 0 ; j < n ; j++){
            if(board[0][j] == 'O' && !vis[0][j]){
                dfs(0 , j , board , vis);
            }
        }

        //last row
        for(int j = 0 ; j < n ; j++){
            if(board[m-1][j] == 'O' && !vis[m-1][j]){
                dfs(m-1 , j , board , vis);
            }
        }

        //1st col
        for(int i = 0 ; i < m ; i++){
            if(board[i][0] == 'O' && !vis[i][0]){
                dfs(i , 0 , board , vis);
            }
        }

        //last col
        for(int i = 0 ; i < m ; i++){
            if(board[i][n-1] == 'O' && !vis[i][n-1]){
                dfs(i , n-1 , board , vis);
            }
        }


        for(int i = 0 ; i < m ; i++){
            for(int j = 0 ; j < n ; j++){
                if(board[i][j] == 'O' && !vis[i][j]){
                    board[i][j] = 'X';
                }
            }
        }

    }

    private void dfs(int i , int j , char[][] board , boolean[][] vis){
        int m = board.length;
        int n = board[0].length;
        if(i < 0 || j < 0 || i >= m || j >= n || board[i][j] != 'O' || vis[i][j] == true ){
            return;
        }

        vis[i][j] = true;
        dfs(i+1 , j , board, vis);
        dfs(i-1 , j , board, vis);
        dfs(i , j+1 , board, vis);
        dfs(i , j-1 , board, vis);
    }
}

File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m = len(board)
        n = len(board[0])

        def dfs(i , j):
            if i < 0 or j < 0 or i >= m or j >= n or board[i][j] != 'O':
                return

            board[i][j] = 'Y'
            
            dfs(i+1, j)
            dfs(i-1, j)
            dfs(i, j+1)
            dfs(i, j-1)


        for i in range(m):
            if board[i][0] == 'O':
                dfs(i , 0)
            if board[i][n-1] == 'O':
                dfs(i, n-1)
        
        for j in range(n):
            if board[0][j] == 'O':
                dfs(0, j)
            if board[m-1][j] == 'O':
                dfs(m-1, j)
        
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'Y':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'
